src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/ManagedIndexCoordinator.kt

```diff
@@ -0,0 +1,338 @@
+        return ThreadPool.Names.MANAGEMENT
+    }
+
+    @Suppress("ReturnCount")
+    override fun clusterChanged(event: ClusterChangedEvent) {
+        if (!isIndexStateManagementEnabled()) return
+
+        if (!event.localNodeMaster()) return
+
+        // TODO: Look into event.isNewCluster, can we return early if true?
+        if (event.isNewCluster) { }
+
+        if (!event.metaDataChanged()) return
+
+        launch { sweepClusterChangedEvent(event) }
+    }
+
+    override fun afterStart() {
+        initBackgroundSweep()
+    }
+
+    override fun beforeStop() {
+        scheduledFullSweep?.cancel()
+    }
+
+    private fun enable() {
+        initBackgroundSweep()
+        indexStateManagementEnabled = true
+    }
+
+    private fun disable() {
+        scheduledFullSweep?.cancel()
+        indexStateManagementEnabled = false
+    }
+
+    private fun isIndexStateManagementEnabled(): Boolean = indexStateManagementEnabled == true
+
+    @OpenForTesting
+    suspend fun sweepClusterChangedEvent(event: ClusterChangedEvent) {
+        val indicesDeletedRequests = event.indicesDeleted()
+                    .filter { event.previousState().metaData().index(it).getPolicyName() != null }
+                    .map { deleteManagedIndexRequest(it.uuid) }
+
+        /*
+        * TODO: Test performance this has on master
+        *
+        * Update existing indices that have added, removed, or updated policy_names
+        * There doesn't seem to be a fast way of finding indices that meet the above conditions without
+        * iterating over every index in cluster state and comparing current policy_name with previous policy_name
+        * If this turns out to be a performance bottle neck we can remove this and enforce
+        * addPolicy/removePolicy API usage for existing indices and let the background sweep pick up
+        * any changes from users that ignore and use the ES settings API
+        * */
+        val requests: List<DocWriteRequest<*>> = event.state().metaData().indices().mapNotNull {
+            val previousIndexMetaData = event.previousState().metaData().index(it.value.index)
+            val policyName = it.value.getPolicyName()
+            var request: DocWriteRequest<*>? = null
+
+            if (it.value.shouldCreateManagedIndexConfig(previousIndexMetaData) && policyName != null) {
+                request = createManagedIndexRequest(it.value.index.name, it.value.indexUUID, policyName)
+            }
+
+            if (it.value.shouldDeleteManagedIndexConfig(previousIndexMetaData)) {
+                // TODO: Hard delete (unsafe) vs delayed delete (safe, but index is still managed)
+                request = deleteManagedIndexRequest(it.value.indexUUID)
+            }
+
+            if (it.value.shouldUpdateManagedIndexConfig(previousIndexMetaData) && policyName != null) {
+                request = updateManagedIndexRequest(ClusterStateManagedIndexConfig(index = it.value.index.name,
```

Comment:
From this `if-statements` it looks like the `request` can be overridden by the last `true` `if-statement`. If the three `if-statements` cannot be overlapping can you use `else-if` or `when` to ensure the request will be unique?